lib/model/CTokenListDataCategorizerBase.cc

```diff
@@ -659,48 +630,99 @@ std::size_t CTokenListDataCategorizerBase::numMatches(CLocalCategoryId categoryI
 }
 
 CDataCategorizer::TLocalCategoryIdVec
-CTokenListDataCategorizerBase::usurpedCategories(CLocalCategoryId categoryId) {
-    CDataCategorizer::TLocalCategoryIdVec usurped;
+CTokenListDataCategorizerBase::usurpedCategories(CLocalCategoryId categoryId) const {
     if (categoryId.isValid() == false || categoryId.index() >= m_Categories.size()) {
         LOG_ERROR(<< "Programmatic error - unexpected ML local category: " << categoryId);
-        return usurped;
+        return {};
     }
     auto iter = std::find_if(m_CategoriesByCount.begin(), m_CategoriesByCount.end(),
                              [categoryId](const TSizeSizePr& pr) {
                                  return pr.second == categoryId.index();
                              });
     if (iter == m_CategoriesByCount.end()) {
-        LOG_WARN(<< "Could not find category definition for category: " << categoryId);
-        return usurped;
+        // If we get here it indicates a bug, as it means m_Categories and
+        // m_CategoriesByCount are inconsistent
+        LOG_ERROR(<< "Programmatic error - ML local category " << categoryId
+                  << " not in categories by count vector");
+        return {};
     }
+    return this->usurpedCategories(iter);
+}
 
-    const CTokenListCategory& category{m_Categories[categoryId.index()]};
+CDataCategorizer::TLocalCategoryIdVec
+CTokenListDataCategorizerBase::usurpedCategories(TSizeSizePrVecCItr iter) const {
+    CDataCategorizer::TLocalCategoryIdVec usurped;
+    if (iter == m_CategoriesByCount.end()) {
+        return usurped;
+    }
```

Comment:
Yes, it's just defensive in case someone calls the function from somewhere else in the future.

Reply:
I generally favour being defensive on all functions inputs for exactly the same reason; although I always feel conflicted about private functions since a reasonable option also seems like check all external input and no internal input. However, I do think we can at least remove the checking and move the log error from `CTokenListDataCategorizerBase::usurpedCategories(CLocalCategoryId categoryId) const` in that case.